src/python/pants/core/goals/package.py

```diff
@@ -31,8 +31,10 @@ class PackageFieldSet(FieldSet, metaclass=ABCMeta):
 @dataclass(frozen=True)
 class BuiltPackage:
     digest: Digest
-    relpath: str
-    extra_log_info: Optional[str] = None
+    relpaths: Tuple[
+        str, ...
+    ]  # Used only for logging. Should reflect the package's paths in digest.
+    extra_logs: Tuple[str, ...] = tuple()
```

Comment:
So the idea of `extra_logs` is that the message will appear in the same `[INFO]` as the "wrote dist to path/" message. I think this is important - when you run `./pants package ::`, you'll get lots of `[INFO]` messages, and it's consistent if each `[INFO]` corresponds to exactly one asset. I'm concerned a user will not realize that the extra logs correspond to a prior `[INFO]`.

This is especially a concern if you're returning multiple assets in one `BuiltPackage`, and the log info applies to each asset. That won't be clear; you'd have multiple `[INFO] wrote to dist/`, and only show the extra info once at the end, making it look like it only applies to the last asset.

--

Instead, thoughts on coupling each relpath with an optional extra log? Ugly strawman design:

```python
relpaths_and_extra_log_infos: Iterable[Tuple[str, Optional[str]]]

[("aws1.zip", "runtime handler: python3.6), ("aws2.zip", "runtime handler: python 3.7")]
```

Reply:
We currently have exactly one use case for the extra log infos, which is logging the runtime and handler info for an AWS Lambda. We don't have the case you're mentioning, and possibly never will. I don't think the nuance of one vs. multiple log lines is really worth the extra hackiness right now.

Reply:
How about simply zipping `relpaths` and `extra_log_infos`? That would achieve what I'm asking for: that the extra info is in the same `[INFO]` as the `wrote dist/` message.

It would avoid the ugly strawman design. The end user API would be the same, only with a new invariant that `relpaths` should align with `extra_logs`

Reply:
That would require checking that the invariant, and is also clunky. 

I don't think this matters much TBH, but in the interest of moving forward I've created a dataclass to hold an artifact path (not "asset" - that was not the right terminology I think, asset is like an image file) plus optional extra log info.

Reply:
On second thought, I still think these actually should be separate log lines. This:

```
$ ./pant package helloworld:helloworld-awslambda
23:16:26.38 [INFO] Wrote dist/helloworld/helloworld-awslambda.zip
23:16:26.38 [INFO]   Runtime: python3.7
23:16:26.38 [INFO]   Handler: lambdex_handler.handler
```

Looks better (wrt indentation) than:

```
$ ./pants package helloworld:helloworld-awslambda
23:13:31.62 [INFO] Wrote dist/helloworld/helloworld-awslambda.zip
  Runtime: python3.7
  Handler: lambdex_handler.handler
```

Reply:
Counterpoint, with streaming results, we use one log per distinct result. I think it's good to be consistent here. 

And then there's the problem of when you have multiple results for the same BuiltPackage, to visually clarify what the extra information corresponds to. 

(Ik this doesn't seem very important, but this is our UX. We've spent lots of time tweaking log messages, and I do think it's worth stressing the details here.)

Reply:
The current implementation has extra information for each artifact, via the BuiltPackageArtifact dataclass, so the multiple results problem is solved. 

I don't like the "single log entry spanning multiple lines" for various reasons, but I can make that change now and we can debate it at leisure.